File: QRCodeScan/QRCodeScan/Main/Controller/ViewController/ViewController.m
//
//  ViewController.m
//  QRCodeScan
//
//  Created by ZHK on 2018/7/26.
//  Copyright © 2018年 ZHK. All rights reserved.
//

#import "ViewController.h"
#import "QRHistoryViewCell.h"
#import "QRResultViewController.h"
#import "WKWebViewController.h"
#import <Realm.h>
#import "QRResult.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, strong) RLMResults *history;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self loadData];
    [_tableView reloadData];
}

#pragma mark - UI

- (void)setupUI {
    self.title = @"扫描历史";
    [_tableView registerClass:[QRHistoryViewCell class] forCellReuseIdentifier:QRHistoryViewCell_IDFR];
}

#pragma mark - Data

- (void)loadData {
    self.history = [[QRResult allObjects] sortedResultsUsingKeyPath:@"date" ascending:NO];
}

#pragma mark - UITableView dataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _history.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    QRHistoryViewCell *cell = [tableView dequeueReusableCellWithIdentifier:QRHistoryViewCell_IDFR];
    cell.QRinfo = _history[indexPath.row];
    return cell;
}

#pragma mark - UITableView delegate

// row select
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    QRResultViewController *resultVC = [[QRResultViewController alloc] init];
    resultVC.res = _history[indexPath.row];
    [self.navigationController pushViewController:resultVC animated:YES];
    
//    QRResult *res = _history[indexPath.row];
//    WKWebViewController *wkwebVC = [[WKWebViewController alloc] init];
//    wkwebVC.url = [NSURL URLWithString:res.url];
//    [self.navigationController pushViewController:wkwebVC animated:YES];
}

// row heighr
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 70.0;
}

// header footer height
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 0.1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    return 0.1;
}

// view for header footer
- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    return nil;
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    return nil;
}

#pragma mark - UITableView edit

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewCellEditingStyleDelete;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        QRResult *res = _history[indexPath.row];
        [res.realm transactionWithBlock:^{
            [res.realm deleteObject:res];
        }];
        [self loadData];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationLeft];
    }
}

@end

File: SyncDemo/SyncDemo/ViewController.swift
//
//  ViewController.swift
//  SyncDemo
//
//  Created by ZHK on 2018/7/23.
//  Copyright © 2018年 ZHK. All rights reserved.
//

import UIKit
import RealmSwift

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    @IBOutlet weak var tableView: UITableView!
    
    var persons: Results<Person>?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.register(UITableViewCell().classForCoder, forCellReuseIdentifier: "cell")
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        loadData()
    }
    
    // MARK:
    
    func loadData() {
        let realm = try! Realm()
        persons = realm.objects(Person.self)
        tableView.reloadData()
    }

    // MARK: TableView dataSource

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return persons?.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell")
        let person = persons![indexPath.row]
        cell?.textLabel?.text = person.name
        return cell!
    }
    
    // MARK: TabelView delegate
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let sb = UIStoryboard(name: "Main", bundle: nil)
        let personVC = sb.instantiateViewController(withIdentifier: "person") as! PersonViewController
        personVC.person = persons?[indexPath.row]
        navigationController?.pushViewController(personVC, animated: true)
    }
}

